### PR TITLE
cpu stream does not have device param

### PR DESCRIPTION
**Pull request by francesco-bertolotti**
## Issue

When using CPU offload with Muon, resharding fails at the following line:

```python
transfer_stream = device_handle.Stream(device=device, priority=0)
```

From what I can gather, `torch.get_device_module(device)` returns `torch.cuda` for CUDA devices and `torch.cpu` when using CPU offload.

The `Stream` implementations for these two modules have different signatures:

* https://github.com/pytorch/pytorch/blob/51aa1bf319ea65e7070d9da5ceb8deb5dc54f600/torch/cuda/streams.py#L37-L46 accepts a `device` argument.
* https://github.com/pytorch/pytorch/blob/51aa1bf319ea65e7070d9da5ceb8deb5dc54f600/torch/cpu/__init__.py#L136-L142 does not accept a `device` argument.

As a result, the current implementation raises an error when running with CPU offload because it passes an unsupported `device` argument to `torch.cpu.Stream`.

## Proposed Fix

Instantiate the `Stream` differently depending on the device type:

* For CPU, create the stream without the `device` argument.
* For GPU, preserve the existing behavior and pass the `device` argument.

## Disclaimer

I am not very familiar with this part of PyTorch. I did some initial experimentation, and the proposed change appears to work.

However, I have not yet performed extensive testing to verify that this is the correct approach or that it behaves properly in all relevant scenarios. I would appreciate a second opinion and feedback on whether this is the appropriate way to handle resharding on cpu

### CODE REVIEW

File: torchtitan/distributed/flex_shard/_optimizer_reshard_runtime.py [hunk 1]
```diff
@@ -264,7 +264,11 @@ def create(
         device: torch.device,
     ) -> _CommunicationContext:
         device_handle = torch.get_device_module(device)
-        transfer_stream = device_handle.Stream(device=device, priority=0)
+        transfer_stream = (
```
Comment (shuhuayu) [inline torchtitan/distributed/flex_shard/_optimizer_reshard_runtime.py]: lgtm, maybe we can add a test with cpu offloading to avoid regression.

Comment (weifengpy) [reply torchtitan/distributed/flex_shard/_optimizer_reshard_runtime.py]: i dont rememeber fsdp2 has such if-else- will look it up when i am around laptop

**Comment by weifengpy** [comment]
@francesco-bertolotti are you seriously looking for cpu-based DistMuon? this is different from running AdamW on cpu. DistMuon has collective and relies on device/cuda stream for comm-compute overlapping. DistMuon on cpu is for debugging purpose in my opinion

edit: this is almost like running fsdp2 with AG/RS on gloo. only for debugging it makes sense

**Comment by francesco-bertolotti** [comment]
> DistMuon is collective and relies on device/CUDA streams for communication-computation overlap. In my opinion, the CPU implementation of DistMuon is mainly intended for debugging.

I have access to Grace Hopper systems, so I have quite a bit of CPU memory available. To fit as many parameters as possible, I’m looking into offloading the master weights and optimizer states to CPU/Grace memory.

I’m also using FlexShard from TorchTitan to ensure that each Muon matrix lands entirely on a single rank. I then stage the Newton-Schulz iterations on the GPU.

With this setup, there shouldn’t really be any communication involved, apart from possibly gradient clipping. So perhaps I shouldn’t be using the distributed implementation for this use case.

@weifengpy, thanks for weighing in on this. What do you think? Is this a use case that DistMuon isn’t really intended to support, or is there a reason I should avoid using it this way?

---

Let me know if you want the test @shuhuayu was mentioning

File: torchtitan/distributed/flex_shard/_optimizer_reshard_runtime.py (hunk 1 above)
Comment (weifengpy) [reply torchtitan/distributed/flex_shard/_optimizer_reshard_runtime.py]: here is how fsdp2 avoids if-else for cpu device

```
-        transfer_stream = device_handle.Stream(device=device, priority=0)
+        with device_handle.stream(device_handle.current_stream(device)):
+            transfer_stream = device_handle.Stream(priority=0)
```

**Review by weifengpy — COMMENTED** [review]
see above

**Comment by weifengpy** [comment]
> What do you think? Is this a use case that DistMuon isn’t really intended to support, or is there a reason I should avoid using it this way?

thanks explaining the context. you know what's going on here then. the fix itself makes sense. I commented inline on how to do it without if-else

**Review by weifengpy — APPROVED** [review]

**Review by tianyu-l — APPROVED** [review]
